File: src/com/mishiranu/dashchan/widget/PostBorderView.java
package com.mishiranu.dashchan.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

import com.mishiranu.dashchan.util.ResourceUtils;
import com.mishiranu.dashchan.widget.ThemeEngine;

public class PostBorderView extends View {
	private BorderStyle borderStyle;
	private BorderConfiguration borderConfiguration;

	public PostBorderView(Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
	}

	public void setBorderStyle(BorderStyle borderStyle) {
		if (this.borderStyle != borderStyle) {
			this.borderStyle = borderStyle;
			initializeBorderConfigurationForStyle();
			requestLayout();
		}
	}

	private void initializeBorderConfigurationForStyle() {
		if (borderStyle == null) {
			borderConfiguration = null;
		} else {
			switch (borderStyle) {
				case USER_POST:
					borderConfiguration = new UserPostBorderConfiguration(getContext());
					break;
				case REPLY:
					borderConfiguration = new ReplyBorderConfiguration(getContext());
					break;
				default:
					borderConfiguration = null;
			}
		}
	}

	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		super.onMeasure(widthMeasureSpec, heightMeasureSpec);
		if (borderConfiguration != null) {
			int borderLeft = getPaddingLeft();
			int borderTop = getPaddingTop();
			int borderRight = getMeasuredWidth() - getPaddingRight();
			int borderBottom = getMeasuredHeight() - getPaddingBottom();
			borderConfiguration.configure(borderLeft, borderTop, borderRight, borderBottom);
		}
	}

	@Override
	protected void onDraw(Canvas canvas) {
		if (borderConfiguration != null) {
			borderConfiguration.draw(canvas);
		}
	}

	public enum BorderStyle {
		USER_POST, REPLY
	}

	private interface BorderConfiguration {
		void configure(int borderLeft, int borderTop, int borderRight, int borderBottom);

		void draw(Canvas canvas);
	}

	private static class UserPostBorderConfiguration implements BorderConfiguration {
		private final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
		private float borderTop;
		private float borderBottom;
		private float borderMid;
		private final float density;

		UserPostBorderConfiguration(Context context) {
			paint.setColor(ThemeEngine.getTheme(context).accent);
			paint.setStrokeCap(Paint.Cap.ROUND);
			density = ResourceUtils.obtainDensity(context);
		}

		@Override
		public void configure(int borderLeft, int borderTop, int borderRight, int borderBottom) {
			float userPostBorderSize = borderRight - borderLeft - density;
			if (userPostBorderSize > 0) {
				paint.setStrokeWidth(userPostBorderSize);

				float borderCapSize = userPostBorderSize / 2;
				this.borderTop = borderTop + borderCapSize;
				this.borderBottom = borderBottom - borderCapSize;
				this.borderMid = (borderLeft + borderRight) / 2f;
			}
		}

		@Override
		public void draw(Canvas canvas) {
			canvas.drawLine(borderMid, borderTop, borderMid, borderBottom, paint);
		}

	}

	private static class ReplyBorderConfiguration implements BorderConfiguration {
		private final Paint paint = new Paint(android.graphics.Paint.ANTI_ALIAS_FLAG);
		private float dotRadius;
		private int dotsCount;
		private float spaceBetweenDots;
		private float dotCenterX;
		private float dotCenterY;

		ReplyBorderConfiguration(Context context) {
			paint.setColor(ThemeEngine.getTheme(context).accent);
		}

		@Override
		public void configure(int borderLeft, int borderTop, int borderRight, int borderBottom) {
			int dotDiameter = borderRight - borderLeft;
			if (dotDiameter > 0) {
				dotRadius = dotDiameter / 2f;

				int availableHeight = borderBottom - borderTop;
				dotsCount = (availableHeight / 2) / dotDiameter;

				spaceBetweenDots = 0;
				if (dotsCount >= 2) {
					int remainingHeight = availableHeight - (dotsCount * dotDiameter);
					spaceBetweenDots += dotDiameter;
					spaceBetweenDots += (float) remainingHeight / (dotsCount - 1); // the last dot does not need a space after it, so divide remaining height between all dots except last
				}

				dotCenterX = dotRadius;
				dotCenterY = borderTop + dotRadius;
			}
		}

		@Override
		public void draw(Canvas canvas) {
			for (int dotNumber = 0; dotNumber < dotsCount; dotNumber++) {
				canvas.drawCircle(dotCenterX, dotCenterY + (spaceBetweenDots * dotNumber), dotRadius, paint);
			}
		}
	}

}

File: src/com/mishiranu/dashchan/widget/PostBorderView.kt
package com.mishiranu.dashchan.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.mishiranu.dashchan.util.ResourceUtils

class PostBorderView(
    context: Context,
    attrs: AttributeSet?,
) : View(context, attrs) {
    private var borderStyle: BorderStyle? = null
    private var borderConfiguration: BorderConfiguration? = null

    fun setBorderStyle(borderStyle: BorderStyle?) {
        if (this.borderStyle != borderStyle) {
            this.borderStyle = borderStyle
            initializeBorderConfigurationForStyle()
            requestLayout()
        }
    }

    private fun initializeBorderConfigurationForStyle() {
        borderConfiguration =
            when (borderStyle) {
                null -> null
                BorderStyle.USER_POST -> UserPostBorderConfiguration(context)
                BorderStyle.REPLY -> ReplyBorderConfiguration(context)
            }
    }

    override fun onMeasure(
        widthMeasureSpec: Int,
        heightMeasureSpec: Int,
    ) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        borderConfiguration?.let {
            val borderLeft = paddingLeft
            val borderTop = paddingTop
            val borderRight = measuredWidth - paddingRight
            val borderBottom = measuredHeight - paddingBottom
            it.configure(borderLeft, borderTop, borderRight, borderBottom)
        }
    }

    override fun onDraw(canvas: Canvas) {
        borderConfiguration?.draw(canvas)
    }

    enum class BorderStyle {
        USER_POST,
        REPLY,
    }

    private interface BorderConfiguration {
        fun configure(
            borderLeft: Int,
            borderTop: Int,
            borderRight: Int,
            borderBottom: Int,
        )

        fun draw(canvas: Canvas)
    }

    private class UserPostBorderConfiguration(
        context: Context,
    ) : BorderConfiguration {
        private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
        private var borderTop = 0f
        private var borderBottom = 0f
        private var borderMid = 0f
        private val density: Float

        init {
            paint.color = ThemeEngine.getTheme(context).accent
            paint.strokeCap = Paint.Cap.ROUND
            density = ResourceUtils.obtainDensity(context)
        }

        override fun configure(
            borderLeft: Int,
            borderTop: Int,
            borderRight: Int,
            borderBottom: Int,
        ) {
            val userPostBorderSize = borderRight - borderLeft - density
            if (userPostBorderSize > 0) {
                paint.strokeWidth = userPostBorderSize

                val borderCapSize = userPostBorderSize / 2
                this.borderTop = borderTop + borderCapSize
                this.borderBottom = borderBottom - borderCapSize
                this.borderMid = (borderLeft + borderRight) / 2f
            }
        }

        override fun draw(canvas: Canvas) {
            canvas.drawLine(borderMid, borderTop, borderMid, borderBottom, paint)
        }
    }

    private class ReplyBorderConfiguration(
        context: Context,
    ) : BorderConfiguration {
        private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
        private var dotRadius = 0f
        private var dotsCount = 0
        private var spaceBetweenDots = 0f
        private var dotCenterX = 0f
        private var dotCenterY = 0f

        init {
            paint.color = ThemeEngine.getTheme(context).accent
        }

        override fun configure(
            borderLeft: Int,
            borderTop: Int,
            borderRight: Int,
            borderBottom: Int,
        ) {
            val dotDiameter = borderRight - borderLeft
            if (dotDiameter > 0) {
                dotRadius = dotDiameter / 2f

                val availableHeight = borderBottom - borderTop
                dotsCount = (availableHeight / 2) / dotDiameter

                spaceBetweenDots = 0f
                if (dotsCount >= 2) {
                    val remainingHeight = availableHeight - dotsCount * dotDiameter
                    spaceBetweenDots += dotDiameter
                    // the last dot does not need a space after it, so divide remaining
                    // height between all dots except last
                    spaceBetweenDots += remainingHeight.toFloat() / (dotsCount - 1)
                }

                dotCenterX = dotRadius
                dotCenterY = borderTop + dotRadius
            }
        }

        override fun draw(canvas: Canvas) {
            for (dotNumber in 0 until dotsCount) {
                canvas.drawCircle(dotCenterX, dotCenterY + spaceBetweenDots * dotNumber, dotRadius, paint)
            }
        }
    }
}
